x/staking/abci.go

```diff
@@ -0,0 +1,43 @@
+package staking
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// BeginBlocker will persist the current header and validator set as a historical entry
+// and prune the oldest entry based on the HistoricalEntries parameter
+func BeginBlocker(ctx sdk.Context, k Keeper) {
+	entryNum := k.HistoricalEntries(ctx)
+	// if there is no need to persist historicalInfo, return
+	if entryNum == 0 {
+		return
+	}
+
+	// Create HistoricalInfo struct
+	lastVals := k.GetLastValidators(ctx)
+	types.Validators(lastVals).Sort()
+	historicalEntry := types.HistoricalInfo{
```

Comment:
we can use the `NewHistoricalInfo` constructor here

Reply:
also, the validation (+ panic on error) is missing

Reply:
Don't know if validation here is necessary. Validation is true by construction, so long as Tendermint consensus is working correctly

Reply:
where is the validation used then?